### ISSUE TITLE
prune dead bunnies

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/natural-selection#112]
To conserve memory, dead bunnies should be pruned (disposed) when they are no longer needed.  And the only need for dead bunnies is for the Pedigree graph.

**Comment by pixelzoom** [comment]
@amanda-phet suggestion in Slack:

![IMG_20200629_170735](https://user-images.githubusercontent.com/3046552/86179525-acf26580-bae7-11ea-97d6-633fa513869f.jpg)

If bunny 0 can't mate with a gen 5 bunny, only a gen 4, then this looks even better and only gets to 15

**Comment by pixelzoom** [comment]
Bunnies die at age 5 _before_ they mate.  So I think the above diagram should probably be for 0 -> 4, correct @amanda-phet?

If that's the case, then we have:

```
0 - 4
  |
  5 - 9
    | 
    10 - 14 
       |
       15
```

**Comment by pixelzoom** [comment]
Bunnies die at age 5, before they mate.  If the depth of the Pedigree tree is 4, can we simply dispose of a bunny when it has been dead for 3 generations?

**Comment by pixelzoom** [comment]
Another problem to solve when disposing of dead bunnies... When a bunny is disposed, it offspring will have a reference to it, via the Bunny `father` or `mother` field.  There's no link from the parent back to its offspring, and we don't want to have to search all bunnies.   In any case, that reference needs to be removed, or the memory won't be freed.   And we can't just set `father` or `mother` to `null`, since that has special meaning (a generation zero bunny).

**Comment by amanda-phet** [comment]
>Bunnies die at age 5, before they mate. If the depth of the Pedigree tree is 4, can we simply dispose of a bunny when it has been dead for 3 generations?

If we dispose of a bunny after it's been dead for 3 generations, that could potentially be a problem for the pedigree. Bunny 0 could have a great-grandchild in generation 15, but not later than that. So, as you said in https://github.com/phetsims/natural-selection/issues/112#issuecomment-653692159, we need to keep bunnies for 15 generations after they are born to handle this worst-case scenario. 

>When a bunny is disposed, it offspring will have a reference to it, via the Bunny father or mother field. There's no link from the parent back to its offspring, and we don't want to have to search all bunnies. In any case, that reference needs to be removed, or the memory won't be freed. And we can't just set father or mother to null, since that has special meaning (a generation zero bunny).

I'm not sure how to handle this particular issue. I'm guessing devs would have more input here than a designer.

**Comment by pixelzoom** [comment]
> I'm not sure how to handle this particular issue. I'm guessing devs would have more input here than a designer.

Yeah, that was more of a note to myself, and has to do with internal implementation. Definitely not a designer issue.

One other note to myself....  Keeping bunnies for 15 generations is still liable to result in a memory problem.  We really only need to keep enough bunnies to satisfy the Pedigree for bunnies that are selectable (alive) plus the bunny that is currently selection (who may be dead or alive).  If I can figure out how to do that, it will require much less memory.  It might require some sort of reference counting.  This is beginning to feel like a full-blown GC problem.

**Comment by pixelzoom** [comment]
7/7/2020 Discussed with @amanda-phet via Zoom.  We'll start by disposing of bunnies who were born 16 generations ago, and see how that affects memory usage.  We can monitor the number of dead bunnies in the browser console using the `?log` query parameter.  For example:

> 2 bunnies died of old age, total dead = 51

**Comment by pixelzoom** [comment]
This is implemented in the above commit.  But I have not been able to test it sufficiently, because I can't get the number of generations to go beyond 7 or 8 before bunnies take over the world. 

I'll leave this open for further testing.  @amanda-phet as in https://github.com/phetsims/natural-selection/issues/46#issuecomment-654993065, if you have a procedure for stabilizing the populate so that the sim runs for many generations, I'm all ears.

**Comment by pixelzoom** [comment]
Uh oh. There's a problem with pruning bunnies 16 generations after they were born.  One of those bunnies could be the selected bunny in the Pedigree graph, and pruning it would cause it to disappear from the graph.  For example, suppose that I select one of the original bunnies before pressing "Add a Mate", and then never select a different bunny.   That original bunny can't be pruned as long as it's selected.  So I think what needs to be done is to modify the algorithm to never prune the selected bunny.

**Comment by pixelzoom** [comment]
I'm also wondering about how I'm currently handling a bunny's references to its parents.  In Bunny.js, a bunny now listens for when its parents have been disposed, and sets `this.father` or `this.mother` to `null`.   But father and mother were previously `null` only for generation 0 bunnies.  Will this cause problems?...

**Comment by pixelzoom** [comment]
To test pruning, I'm also going to add a query parameter that prevents bunnies from taking over the world, as described in https://github.com/phetsims/natural-selection/issues/46#issuecomment-656274546.

**Comment by pixelzoom** [comment]
The above commit moves `selectBunnyProperty` from PedigreeModel to BunnyCollection,  and ensures that the selected bunny is not pruned.  In retrospect, it was probably OK to prune the selected bunny (and its ancestors, which I didn't deal with), because the Pedigree view is static after a bunny has died.   The genotype abbreviation may be updated if a new mutation is applied, but that info is still present for a disposed bunny.   If any disposed bunnies are shown in the Pedigree, their memory won't be freed until the selected bunny changes, but that's OK.

**Comment by pixelzoom** [comment]
I think I'm OK with setting `this.father` or `this.mother` to null when they are disposed.   I looked through the implementation and I don't see any problem.  And setting it to something other than null would complicate the code that constructs the Pedigree graph.  So I've removed the TODO in the above commit.

**Comment by pixelzoom** [comment]
What remains to be done here is....

Use the procedure in https://github.com/phetsims/natural-selection/issues/119#issuecomment-657915594 to put the sim in a stable state:
> 1. Add a mate
> 2. At the start of generation 3, check "Limited Food"

Let the sim run for a large number of generations.  Take heap snapshots to identify memory consumption.

**Comment by pixelzoom** [comment]
This algorithm is working nicely.  Bunnies are being pruned as expected, and memory consumption is stable.

I did the following few times:

1. Run the sim in Chrome, with `?ea&log&secondsPerGeneration=1`.
2. Go to Lab screen
3. Press "Add a Mate" button
4. Check "Limited Food" at the start of Generation 3
5. Take a Heap Snapshot every 100 generations.

The snapshots were all in the range [32,39] MB, as the number of bunnies oscillated between 5 and 25.  And the log showed that the number of dead bunnies was managed via this pruning algorithm, for example:

```
====== Generation 791 ======
live bunnies = 4
dead bunnies = 112
recessive mutants = 0
0 bunnies died of old age
8 dead bunnies pruned
```

Waiting a little later to check "Limited Food" yielded a larger population of bunnies, and the Heap Snapshots were in the 60MB range.  

I also paused the sim every once in awhile to select bunnies in the Pedigree graph, to spot-check that dead bunnies required by a pedigree weren't missing.  No problems noted, but this is more difficult to test rigorously.  (But the code is robust, and will handle a situation where a dead bunny is missing, so the sim will not crash.)

So...   Nothing else to do here, closing.
